Intrinsics JsonToString and StringToJson

kbrock opened the pull request:
part of #64

Not sure if we want to do a special json error parse case

Fryguy commented on lib/floe/workflow/intrinsic_function/transformer.rb:
```diff
@@ -151,6 +151,11 @@ def check_types(args, function, signature)
 
+        rule(:states_json_to_string => {:args => subtree(:args)}) do
+          args = Transformer.process_args(args(), "States.JsonToString", [Object])
+          JSON.generate(args.first)
```
I know it's pedantic, but for code consistency purposes
```suggestion
          json = args.first

          JSON.generate(json)
```

Fryguy replied:
Is it possible for JSON.generate to fail?  It feels like it could, but I'm not sure how.  If it can fail with a specific error, we will likely need to rescue and wrap that in a Floe error.

Fryguy commented on lib/floe/workflow/intrinsic_function/transformer.rb:
```diff
@@ -182,6 +187,11 @@ def check_types(args, function, signature)
 
+        rule(:states_string_to_json => {:args => subtree(:args)}) do
+          args = Transformer.process_args(args(), "States.StringToJson", [String])
+          JSON.parse(args.first)
```
Similarly here:

```suggestion
          string = args.first

          JSON.parse(string)
```

Fryguy commented on spec/workflow/intrinsic_function_spec.rb:
```diff
@@ -450,6 +450,13 @@
+      it "fetches values from the input" do
+        result = described_class.value("States.JsonToString($.input)", {"context" => {"baz" => "qux"}}, {"input" => {"foo" => "bar"}})
+        expect(result).to eq("{\"foo\":\"bar\"}")
+      end
```
Please include a test for the various failure cases (wrong type, too few params, too many params).

Fryguy commented on spec/workflow/intrinsic_function_spec.rb:
```diff
@@ -450,6 +450,13 @@
 
+    describe "States.JsonToString" do
+      it "fetches values from the input" do
+        result = described_class.value("States.JsonToString($.input)", {"context" => {"baz" => "qux"}}, {"input" => {"foo" => "bar"}})
```
The context is not needed for the test, so it visually compicates the test, so let's get rid of it and replace with a hardcoded `{}` like we do in other specs.

Fryguy commented on spec/workflow/intrinsic_function_spec.rb:
```diff
@@ -610,6 +617,13 @@
+      it "fetches values from the input" do
+        result = described_class.value("States.StringToJson($.input)", {"context" => {"baz" => "qux"}}, {"input" => "{\"foo\":\"bar\"}"})
+        expect(result).to eq({"foo" => "bar"})
+      end
```
Please also add the common failure cases as before.

Addtionally, there can be "bad" strings, so let's have specs for those as well (and on that note, they will probably raise a JSON::ParserError that we probably need to capture and convert to a Floe error.  Offhand, this should non-json strings (e.g. "XXX"), the empty string, a string with only a newline, and non-hash JSON strings like Array, Integer, "null", etc (there's probably more we can verify).

Fryguy commented on spec/workflow/intrinsic_function_spec.rb:
```diff
@@ -610,6 +617,13 @@
 
+    describe "States.StringToJson" do
+      it "fetches values from the input" do
+        result = described_class.value("States.StringToJson($.input)", {"context" => {"baz" => "qux"}}, {"input" => "{\"foo\":\"bar\"}"})
```
Similarly please drop the context.

Fryguy commented on lib/floe/workflow/intrinsic_function/transformer.rb:
```diff
@@ -151,6 +151,11 @@ def check_types(args, function, signature)
         end
 
+        rule(:states_json_to_string => {:args => subtree(:args)}) do
+          args = Transformer.process_args(args(), "States.JsonToString", [Object])
```
If it's Object then we should have a separate test for each type of object possible that can be converted to JSON (might be useful to run this through the stepfunctions simulartor to see what it does - my script in #244 is helpful for this purpose)

Fryguy commented on lib/floe/workflow/intrinsic_function/parser.rb:
```diff
@@ -64,9 +64,11 @@ class Parser < Parslet::Parser
           :states_base64_encode,   "States.Base64Encode",
           :states_base64_decode,   "States.Base64Decode",
           :states_hash,            "States.Hash",
+          :states_json_to_string,  "States.JsonToString",
```
This order is not alphabetical - it currently matches the order as listed in the stateslanguage.net spec.  The same goes for the order of the methods within the transformer.

Fryguy replied:
usually for the happy path test I just use the one example from states-language.net :)

Fryguy commented:
This is also missing parser_spec tests - they are kind of dumb - you can copy another one like States.Array - it's basically just checking a happy path, and then the bad path is when someone puts `StringToJson` without parens.

kbrock commented:
update:
- Rebase.
- Add args parameter names to transformer.
- Add bad argument tests.
- Added all class input for JsonToString.
- Drop unneeded contexts from tests.
- Changed order of transformer to match specifications.

Fryguy reviewed (approved):
There a few organizational things that don't match the other specs, but that can all be followups for cleanup, so merging.

Overall, this is great! Thanks @kbrock
